[Event Hubs] Update samples, README, changelog and JSDocs (#4675)
Azure/azure-sdk-for-js: sdk/eventhub/event-hubs/README.md

- [Inspect an Event Hub](#inspect-an-event-hub)
- [Publish events to an Event Hub](#publish-events-to-an-event-hub)
- [Consume events from an Event Hub partition](#consume-events-from-an-event-hub-partition)

> - [Consume events from all partitions of an Event Hub](#consume-events-from-all-partitions-of-an-event-hub)

Reviewer: How about `Consume events using an Event Processor` ?

That will give an idea that this sample is very much different from the previous one

**Note**: When working with Azure Stream Analytics, the body of the event being sent should be a JSON object as well.
For example: `body: { "message": "Hello World" }`

### Consume events from an Event Hub partition

> To consume events for an Event Hub partition, create an `EventHubConsumer` for that partition and consumer group combination. When an Event Hub is created, it provides a default consumer group that can be used to get started. A consumer also needs to specify where in the event stream to begin receiving events; in our example, we will read new events as they are published.

Reviewer: ```suggestion
To consume events from a single Event Hub partition in a consumer group, create an `EventHubConsumer` for that partition and consumer group combination. You will need to provide a position in the event stream from where to begin receiving events; in our example, we will read new events as they are published.
```

> In order to consume events for an Event Hub partition, you'll need to create an `EventHubConsumer` for that partition and consumer group combination. When an Event Hub is created, it provides a default consumer group that can be used to get started. A consumer also needs to specify where in the event stream to begin receiving events; in our example, we will focus on reading new events as they are published.

Reviewer 2: ```suggestion
To consume events for an Event Hub partition, create an `EventHubConsumer` for that partition and consumer group combination. When an Event Hub is created, it provides a default consumer group that can be used to get started. A consumer also needs to specify where in the event stream to begin receiving events; in our example, we will read new events as they are published.
```

}
```

### Consume events using an Event Processor

> To consume events for all partitions of an Event Hub, you'll create an `EventProcessor` for a specific consumer group. When an Event Hub is created, it provides a default consumer group that can be used to get started.

Reviewer 2: ```suggestion
To consume events for all partitions of an Event Hub, create an `EventProcessor` for a specific consumer group. When an Event Hub is created, it provides a default consumer group that can be used to get started.
```

> The `EventProcessor` will delegate processing of events to a `PartitionProcessor` implementation that you provide, allowing your logic to focus on the logic needed to provide value while the processor holds responsibility for managing the underlying consumer operations. In our example, we will focus on building the `EventProcessor` and use a very minimal partition processor that does no actual processing.

Reviewer 2: ```suggestion
The `EventProcessor` will delegate the processing of events to a `PartitionProcessor` that you provide, allowing you to focus on business logic while the processor holds responsibility for managing the underlying consumer operations. In our example, we will focus on building the `EventProcessor` and use a very minimal partition processor that does no actual processing.
```

```javascript
class SimplePartitionProcessor {
  async processEvents(events: ReceivedEventData[]) {
    // your code here
  }

  async processError(error: Error) {
    // your error handler here
  }
}

const processor = new EventProcessor(
  EventHubClient.defaultConsumerGroupName,
  client,
  () => {
    return new SimplePartitionProcessor();

>   },

Reviewer: Lets we collapse the above 3 lines to 1

Author: This is because of formatter. I'll change to 
```javascript
const processor = new EventProcessor(
  EventHubClient.defaultConsumerGroupName,
  client,
  () => new SimplePartitionProcessor(),
  new InMemoryPartitionManager()
);
```

  new InMemoryPartitionManager()
);
await processor.start();
// At this point, the processor is consuming events from each partition of the Event Hub and
// delegating them to the SimplePartitionProcessor instance created for that partition.  This
// processing takes place in the background and will not block.
//
// It is important to note that the processor does not own the EventHubClient that was passed into it.
// You are responsible for ensuring that it is disposed after processing.  It is also important that it

> // not be closed or disposed during the time that the EventProcessor is running.

Reviewer: Can we pull this note on EventHubClient out of the code sample? It doesnt feel like the right place for it

}
```

### Consume events from all partitions of an Event Hub

To consume events for all partitions of an Event Hub, you'll create an `EventProcessor` for a specific consumer group. When an Event Hub is created, it provides a default consumer group that can be used to get started.

> The `EventProcessor` will delegate processing of events to a `PartitionManager` implementation that you provide, allowing your logic to focus on the logic needed to provide value while the processor holds responsibility for managing the underlying consumer operations. In our example, we will focus on building the `EventProcessor` and use a very minimal partition processor that does no actual processing.

Reviewer 3: > The `EventProcessor` will delegate processing of events to a `PartitionManager` implementation
should be
> The `EventProcessor` will delegate processing of events to a `PartitionProcessor` implementation

Author: Ohh Yess!!

> The `EventProcessor` will delegate processing of events to a `PartitionManager` implementation that you provide, allowing your logic to focus on the logic needed to provide value while the processor holds responsibility for managing the underlying consumer operations. In our example, we will focus on building the `EventProcessor` and use a very minimal partition processor that does no actual processing.

Reviewer 3: >  , allowing your logic to focus on the logic needed to provide value while the processor holds responsibility for managing the underlying consumer operations.
The use of double-logic seems a bit weird, what about...
> . The `PartitionProcessor` handles processing events for a single partition, and `EventProcessor` manages the creation and assignment of `PartitionProcessor` instances to Event Hub partitions.

> The `EventProcessor` will delegate processing of events to a `PartitionManager` implementation that you provide, allowing your logic to focus on the logic needed to provide value while the processor holds responsibility for managing the underlying consumer operations. In our example, we will focus on building the `EventProcessor` and use a very minimal partition processor that does no actual processing.

Reviewer 3: Can we also add something about the checkpointing functionality?

> ```javascript

Reviewer: Couple of points to simplify the sample here...

- Update `SimplePartitionProcessor` to only implement the required fields i.e we can skip initialize and close
- Skip passing options to `EventProcessor` constructor
- Simplify the factory to a single line  `() => { return new SimplePartitionProcessor() }`
- Skip the const for the factory and use the definition directly in the constructor

Reviewer: After the sample, you can add some text like

> To control the number of events passed to `processEvents`, use the options argument in the `EventProcessor` constructor

class SimplePartitionProcessor {
  async processEvents(events: ReceivedEventData[]) {

>     // your code here

Reviewer 3: It would be nice to show how to checkpoint as well.

Reviewer: I am torn between keeping the sample simple and showing off all the features...

  }

  async processError(error: Error) {
    // your error handler here
  }

  async initialize() {
    // your code here
  }

  async close() {
    // your code here

>   }

Reviewer: For the purposes of the sample, we might want to skip the optional methods here like` initialize` and `close`
